oligo_designer_toolsuite/oligo_specificity_filter/_filter_base.py

```diff
@@ -298,10 +364,130 @@ def _run_filter(
             consider_hits_from_input_region=consider_hits_from_input_region,
         )
+        if self.ai_filter_api is not None:
+            table_hits = self._ai_filter_hits(
+                sequence_type=sequence_type,
+                table_hits=table_hits,
+                file_reference=file_reference,
+                oligo_database=oligo_database,
+                region_id=region_id,
+            )
+
+        return table_hits
+
+    def _ai_filter_hits(
+        self,
+        sequence_type: _TYPES_SEQ,
+        table_hits: pd.DataFrame,
+        file_reference: str,
+        oligo_database: OligoDatabase,
+        region_id: str,
+    ) -> pd.DataFrame:
+        """Filters the hits from a search operation using Machine Learning models. The Hits that recieve a score form the machine learning model lower that the given threshold are filtered out.
+
+        :param sequence_type: The type of sequences being filtered, must be one of the predefined sequence types.
+        :type sequence_type: _TYPES_SEQ
+        :param table_hits: Dataframe containing the strue hits of the search results.
+        :type table_hits: pd.DataFrame
+        :param file_reference: Path to the fasta file used as reference for the search.
+        :type file_reference: str
+        :param oligo_database: The oligo database to which the filter will be applied.
+        :type oligo_database: OligoDatabase
+        :param region_id: The identifier for the region within the database to filter.
+        :type region_id: str
+        :return: Dataframe containing the filtered true hits
+        :rtype: pd.DataFrame
+        """
+
+        required_fields = [
+            
+        ]
+        # check if there are any oligos to filter
+        if len(table_hits) == 0:
+            return table_hits
+        
+        # generate the references and queries sequences
+        references = self._get_references(table_hits, file_reference, region_id)
+        queries = self._get_queries(sequence_type, table_hits, oligo_database, region_id)
+        assert len(queries) == len(
+            references
+        ), "The reference sequences haven't been correctly retrieved."
+
+        # align the references and queries by adding gaps
+        gapped_queries, gapped_references = self._add_alignement_gaps(
+            table_hits=table_hits, queries=queries, references=references
+        )
+
+        # predict the scores for each hit
+        predictions = self.ai_filter_api.predict(
+            queries=queries,
+            gapped_queries=gapped_queries,
+            references=references,
+            gapped_references=gapped_references,
+        )
+
+        # filter the database, keep only the oligos above the threshold
+        table_hits.reset_index(drop=True, inplace=True)
+        ids_vanilla = len(table_hits["query"].unique())
+        below_threshold = np.where(predictions < self.ai_filter_threshold)[0]
+        table_hits.drop(index=below_threshold, inplace=True)
```

Comment:
Consider optimizing the code above:
table_hits = table_hits[predictions >= self.ai_filter_threshold]